### PR TITLE
fix coca training

### LINKED ISSUE #715: coca training doesn't work
I made a PR that should fix it #710

### PR DESCRIPTION
**Pull request by gpucce**
@rwightman I was too hasty with the `embed_cls` thing, I think now the model would not train. 
This should make it work with the tokenizer and train ok, fix #715

### CODE REVIEW

**Comment by rom1504** [comment]
Would be good to add coca to the training test. That way we'll catch any
future issue

Did this current issue make training fail completely or be wrong ? If it
made it fail completely, all the test has to do is to run the training,
which is what the current test does, just need to include coca there

**Comment by gpucce** [comment]
I think coca is in the training tests already, but this does not error. The tokens are not shifted by one, so it learns to copy current token instead of predict the next one

**Comment by rwightman** [comment]
So, been thinking about this one, I really don't like the is_training, it's not done this way elsewhere. The label shift is standard, but why do we need to truncate the text encoder output like that only for training?

**Comment by gpucce** [comment]
> So, been thinking about this one, I really don't like the is_training, it's not done this way elsewhere. The label shift is standard, but why do we need to truncate the text encoder output like that only for training?


@rwightman 

After shortening the labels one needs to drop the last token in the encoder otherwise there will be a length mismatch and also the last token does have a `next_one` to use as a label. However in generation one wants to keep also the last token otherwise it does not go forward.

One could also do it something like this `self.encode_text(text[:, :-1])`

**Comment by rwightman** [comment]
Yeah I don't like embed_cls either. Truncating the text input first, outside of the forward ala `self.encode_text(text[:, :-1])` is the 'normal' approach, but wasn't sure if that would impact the contrastive latent?

**Comment by gpucce** [comment]
The reason for this is that it was meant to keep it identical to how it was before (assuming I did it right) and since compared to before the tokenizer has a hidden `text[:, :-1]` this way the change would not show in the contrastive latent but be there in the generative logits. 

However, it probably makes very little difference and the 'normal' way is better.

**Comment by rwightman** [comment]
merged through #877 with minor changes
